crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3761,31 +3762,31 @@ impl<'db> TypeInferenceBuilder<'db> {
 impl<'db> TypeInferenceBuilder<'db> {
     fn infer_annotation_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
         // https://typing.readthedocs.io/en/latest/spec/annotations.html#grammar-token-expression-grammar-annotation_expression
-        match expression {
+        let annotation_ty = match expression {
             // TODO: parse the expression and check whether it is a string annotation, since they
             // can be annotation expressions distinct from type expressions.
             // https://typing.readthedocs.io/en/latest/spec/annotations.html#string-annotations
-            ast::Expr::StringLiteral(_literal) => {
-                self.store_expression_type(expression, Type::Todo);
-                Type::Todo
-            }
+            ast::Expr::StringLiteral(_literal) => Type::Todo,
 
             // Annotation expressions also get special handling for `*args` and `**kwargs`.
             ast::Expr::Starred(starred) => self.infer_starred_expression(starred),
 
             // All other annotation expressions are (possibly) valid type expressions, so handle
             // them there instead.
-            type_expr => self.infer_type_expression(type_expr),
-        }
+            type_expr => self.infer_expression_type_no_store(type_expr),
+        };
+
+        self.store_expression_type(expression, annotation_ty);
+        annotation_ty
     }
 }
 
 /// Type expressions
 impl<'db> TypeInferenceBuilder<'db> {
-    fn infer_type_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
+    fn infer_expression_type_no_store(&mut self, expression: &ast::Expr) -> Type<'db> {
```

Comment:
Ugh I'm sorry, this was my carelessness in my review comment, but this should really be `infer_type_expression_no_store`, not `infer_expression_type_no_store`

Reply:
I was careless — sorry. Didn't even look at the diff again :see_no_evil: